### ISSUE TITLE
salt-bootstrap fails to install on Redhat when EPEL repos are not available

### ISSUE THREAD

**Issue by Diggs27** [saltstack/salt-bootstrap#1432]
### Description of Issue/Question
salt-bootstrap fails on Redhat7 when there no subscription is installed and therefore can't get access to the EPEL repo. The issue appears to be related to the libyaml-0.1.4-11.el7_0.x86_64 in the salt stack redhat repository. If the yum repo system can't find the libyaml in its own base repository (due to the repo setup yet) then yum tries to pull the dependency from salt stack repo. When this happens, the package fails to install with error reported below. 

The GPG keys listed for the "SaltStack latest Release Channel for RHEL/CentOS 7" repository are already installed but they are not correct for this package.
Check that the correct key URLs are configured for this repository.

 Failing package is: libyaml-0.1.4-11.el7_0.x86_64
 GPG Keys are configured as: https://repo.saltstack.com/yum/redhat/7/x86_64/latest/SALTSTACK-GPG-KEY.pub

### Setup
Although this issue primarly affects the use case of Redhat7, it can be recreated on Centos7

1. Setup Vagrant file
Vagrant.configure("2") do |config|
  config.vm.box = "centos/7"
end
2. Bring up Centos machine
vagrant up
3. Login as root user
4. Remove all of the Fedora repo packages to force Salt to grab all dependencies from salt stack repo.

### Steps to Reproduce Issue
(Include debug logs if possible, `bootstrap-salt.sh -D`.)

curl https://bootstrap.saltstack.com -o install.sh 
[root@localhost yum.repos.d]# sh install.sh -D
 *  INFO: Running version: 2020.02.04
 *  INFO: Executed by: sh
 *  INFO: Command line: 'install.sh -D'

 *  INFO: System Information:
 *  INFO:   CPU:          GenuineIntel
 *  INFO:   CPU Arch:     x86_64
 *  INFO:   OS Name:      Linux
 *  INFO:   OS Version:   3.10.0-957.12.2.el7.x86_64
 *  INFO:   Distribution: CentOS 7.6

 * DEBUG: Binaries will be searched using the following $PATH: /usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/root/bin
 *  INFO: Installing minion
 * DEBUG: install_centos_7_stable_deps not found....
 * DEBUG: install_centos_7_6_stable_deps not found....
 * DEBUG: install_centos_7_deps not found....
 * DEBUG: install_centos_7_6_deps not found....
 *  INFO: Found function install_centos_stable_deps
 * DEBUG: DEPS_INSTALL_FUNC=install_centos_stable_deps
 * DEBUG: config_centos_7_stable_salt not found....
 * DEBUG: config_centos_7_6_stable_salt not found....
 * DEBUG: config_centos_7_salt not found....
 * DEBUG: config_centos_7_6_salt not found....
 * DEBUG: config_centos_stable_salt not found....
 * DEBUG: config_centos_salt not found....
 *  INFO: Found function config_salt
 * DEBUG: CONFIG_SALT_FUNC=config_salt
 * DEBUG: preseed_centos_7_stable_master not found....
 * DEBUG: preseed_centos_7_6_stable_master not found....
 * DEBUG: preseed_centos_7_master not found....
 * DEBUG: preseed_centos_7_6_master not found....
 * DEBUG: preseed_centos_stable_master not found....
 * DEBUG: preseed_centos_master not found....
 *  INFO: Found function preseed_master
 * DEBUG: PRESEED_MASTER_FUNC=preseed_master
 * DEBUG: install_centos_7_stable not found....
 * DEBUG: install_centos_7_6_stable not found....
 *  INFO: Found function install_centos_stable
 * DEBUG: INSTALL_FUNC=install_centos_stable
 * DEBUG: install_centos_7_stable_post not found....
 * DEBUG: install_centos_7_6_stable_post not found....
 * DEBUG: install_centos_7_post not found....
 * DEBUG: install_centos_7_6_post not found....
 *  INFO: Found function install_centos_stable_post
 * DEBUG: POST_INSTALL_FUNC=install_centos_stable_post
 * DEBUG: install_centos_7_stable_restart_daemons not found....
 * DEBUG: install_centos_7_6_stable_restart_daemons not found....
 * DEBUG: install_centos_7_restart_daemons not found....
 * DEBUG: install_centos_7_6_restart_daemons not found....
 * DEBUG: install_centos_stable_restart_daemons not found....
 *  INFO: Found function install_centos_restart_daemons
 * DEBUG: STARTDAEMONS_INSTALL_FUNC=install_centos_restart_daemons
 * DEBUG: daemons_running_centos_7_stable not found....
 * DEBUG: daemons_running_centos_7_6_stable not found....
 * DEBUG: daemons_running_centos_7 not found....
 * DEBUG: daemons_running_centos_7_6 not found....
 * DEBUG: daemons_running_centos_stable not found....
 * DEBUG: daemons_running_centos not found....
 *  INFO: Found function daemons_running
 * DEBUG: DAEMONS_RUNNING_FUNC=daemons_running
 * DEBUG: install_centos_7_stable_check_services not found....
 * DEBUG: install_centos_7_6_stable_check_services not found....
 * DEBUG: install_centos_7_check_services not found....
 * DEBUG: install_centos_7_6_check_services not found....
 * DEBUG: install_centos_stable_check_services not found....
 *  INFO: Found function install_centos_check_services
 * DEBUG: CHECK_SERVICES_FUNC=install_centos_check_services
 *  INFO: Running install_centos_stable_deps()
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
 * epel: mirror.math.princeton.edu
Package matching yum-utils-1.1.31-29.el7.noarch already installed. Checking for update.
Resolving Dependencies
--> Running transaction check
---> Package PyYAML.x86_64 0:3.11-1.el7 will be installed
--> Processing Dependency: libyaml-0.so.2()(64bit) for package: PyYAML-3.11-1.el7.x86_64
--> Running transaction check
---> Package libyaml.x86_64 0:0.1.4-11.el7_0 will be installed
--> Finished Dependency Resolution

Dependencies Resolved

================================================================================
 Package         Arch           Version                 Repository         Size
================================================================================
Installing:
 PyYAML          x86_64         3.11-1.el7              saltstack         160 k
Installing for dependencies:
 libyaml         x86_64         0.1.4-11.el7_0          saltstack          55 k

Transaction Summary
================================================================================
Install  1 Package (+1 Dependent package)

Total size: 215 k
Installed size: 780 k
Downloading packages:
warning: /var/cache/yum/x86_64/7/saltstack/packages/libyaml-0.1.4-11.el7_0.x86_64.rpm: Header V3 RSA/SHA256 Signature, key ID f4a80eb5: NOKEY
Retrieving key from https://repo.saltstack.com/yum/redhat/7/x86_64/latest/SALTSTACK-GPG-KEY.pub

The GPG keys listed for the "SaltStack latest Release Channel for RHEL/CentOS 7" repository are already installed but they are not correct for this package.
Check that the correct key URLs are configured for this repository.

 Failing package is: libyaml-0.1.4-11.el7_0.x86_64
 GPG Keys are configured as: https://repo.saltstack.com/yum/redhat/7/x86_64/latest/SALTSTACK-GPG-KEY.pub

 * ERROR: Failed to run install_centos_stable_deps()!!!
 * DEBUG: Removing the logging pipe /tmp/bootstrap-salt.logpipe
 * DEBUG: Removing the temporary apt error file /tmp/apt_error.VLIhZ8

### Versions and Systems
(`salt --versions-report`, `bootstrap-salt.sh -v`, system type and version,
cloud/VM provider as appropriate.)
install.sh -- Version 2020.02.0

**Comment by bryceml** [comment]
@Diggs27 I believe this works with the latest version of salt now.  Can you confirm so we can close this issue?

**Comment by Diggs27** [comment]
Attempted to run again with the same process as above and received the following:

[root@localhost ~]# sh install.sh -D
 *  INFO: Running version: 2020.06.23
 *  INFO: Executed by: sh
 *  INFO: Command line: 'install.sh -D'

 *  INFO: System Information:
 *  INFO:   CPU:          GenuineIntel
 *  INFO:   CPU Arch:     x86_64
 *  INFO:   OS Name:      Linux
 *  INFO:   OS Version:   3.10.0-1127.el7.x86_64
 *  INFO:   Distribution: CentOS 7.8

 * DEBUG: Binaries will be searched using the following $PATH: /usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/root/bin
 *  INFO: Installing minion
 * DEBUG: install_centos_7_stable_deps not found....
 * DEBUG: install_centos_7_8_stable_deps not found....
 * DEBUG: install_centos_7_deps not found....
 * DEBUG: install_centos_7_8_deps not found....
 *  INFO: Found function install_centos_stable_deps
 * DEBUG: DEPS_INSTALL_FUNC=install_centos_stable_deps
 * DEBUG: config_centos_7_stable_salt not found....
 * DEBUG: config_centos_7_8_stable_salt not found....
 * DEBUG: config_centos_7_salt not found....
 * DEBUG: config_centos_7_8_salt not found....
 * DEBUG: config_centos_stable_salt not found....
 * DEBUG: config_centos_salt not found....
 *  INFO: Found function config_salt
 * DEBUG: CONFIG_SALT_FUNC=config_salt
 * DEBUG: preseed_centos_7_stable_master not found....
 * DEBUG: preseed_centos_7_8_stable_master not found....
 * DEBUG: preseed_centos_7_master not found....
 * DEBUG: preseed_centos_7_8_master not found....
 * DEBUG: preseed_centos_stable_master not found....
 * DEBUG: preseed_centos_master not found....
 *  INFO: Found function preseed_master
 * DEBUG: PRESEED_MASTER_FUNC=preseed_master
 * DEBUG: install_centos_7_stable not found....
 * DEBUG: install_centos_7_8_stable not found....
 *  INFO: Found function install_centos_stable
 * DEBUG: INSTALL_FUNC=install_centos_stable
 * DEBUG: install_centos_7_stable_post not found....
 * DEBUG: install_centos_7_8_stable_post not found....
 * DEBUG: install_centos_7_post not found....
 * DEBUG: install_centos_7_8_post not found....
 *  INFO: Found function install_centos_stable_post
 * DEBUG: POST_INSTALL_FUNC=install_centos_stable_post
 * DEBUG: install_centos_7_stable_restart_daemons not found....
 * DEBUG: install_centos_7_8_stable_restart_daemons not found....
 * DEBUG: install_centos_7_restart_daemons not found....
 * DEBUG: install_centos_7_8_restart_daemons not found....
 * DEBUG: install_centos_stable_restart_daemons not found....
 *  INFO: Found function install_centos_restart_daemons
 * DEBUG: STARTDAEMONS_INSTALL_FUNC=install_centos_restart_daemons
 * DEBUG: daemons_running_centos_7_stable not found....
 * DEBUG: daemons_running_centos_7_8_stable not found....
 * DEBUG: daemons_running_centos_7 not found....
 * DEBUG: daemons_running_centos_7_8 not found....
 * DEBUG: daemons_running_centos_stable not found....
 * DEBUG: daemons_running_centos not found....
 *  INFO: Found function daemons_running
 * DEBUG: DAEMONS_RUNNING_FUNC=daemons_running
 * DEBUG: install_centos_7_stable_check_services not found....
 * DEBUG: install_centos_7_8_stable_check_services not found....
 * DEBUG: install_centos_7_check_services not found....
 * DEBUG: install_centos_7_8_check_services not found....
 * DEBUG: install_centos_stable_check_services not found....
 *  INFO: Found function install_centos_check_services
 * DEBUG: CHECK_SERVICES_FUNC=install_centos_check_services
 *  INFO: Running install_centos_stable_deps()
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
 * epel: dl.fedoraproject.org
Resolving Dependencies
--> Running transaction check
---> Package PyYAML.x86_64 0:3.11-1.el7 will be installed
--> Processing Dependency: libyaml-0.so.2()(64bit) for package: PyYAML-3.11-1.el7.x86_64
--> Finished Dependency Resolution
Error: Package: PyYAML-3.11-1.el7.x86_64 (saltstack)
           Requires: libyaml-0.so.2()(64bit)
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest
 * ERROR: Failed to run install_centos_stable_deps()!!!
 * DEBUG: Removing the logging pipe /tmp/bootstrap-salt.logpipe
 * DEBUG: Removing the temporary apt error file /tmp/apt_error.McIiNl

**Comment by bryceml** [comment]
It appears the missing packages are pciutils and systemd-python which don't appear to be in epel anyway.  I don't think installing without the redhat repos is a supported thing though.  @dmurphy18 can you confirm that installing salt without the redhat repos available is not supported?

**Comment by dmurphy18** [comment]
Essentially Salt expects to be able to load dependent packages from the Operating System, in the case of Redhat / Centos without the need for EPEL.  If those dependencies are not available from the OS then the install will fail.  

For example: on Redhat / Centos, if someone removed installed packages since they thought they no longer needed them and did not have the install media available, then Salt would fail to install.  This case actually happened and is why there is a base sub-directory containing some dependencies that had been removed, url http://repo.saltstack.com/py3/redhat/7/x86_64/latest/base/

And while some of the specified dependencies are included in that directory, systemd-python is not.  This was a choice since at the time, rebuilding systemd-python was large, complex and tied to versions of systemd, and it was felt that any system missing systemd-python had bigger issues too.

Hence in order to install in this case, the needed dependencies need to be available from the OS.

Note that the issue of the GPG Key is due to not leveraging the install rpm from Salt, http://repo.saltstack.com/py3/redhat/salt-py3-repo-latest.el7.noarch.rpm

This installs not only the SaltStack Public GPG key but also the Centos 7 Public GPG Key which has signed http://repo.saltstack.com/py3/redhat/7/x86_64/latest/base/libyaml-0.1.4-11.el7_0.x86_64.rpm

Closing this since installing Redhat from bootstrap needs to have dependencies available to Salt, an alternate would be to use Centos which does not require a subscription manager.  Installing the Centos 7 Public GPG Key  on the system would allow for libyaml-0.1.4-11.el7_0.x86_64.rpm to be installed from SaltStack but without systemd-python available the install would still fail.

If there are other considerations for discussion with this issue, please reopen it and draw attention to it.
